cookbooks/fb_networkmanager/spec/default_spec.rb

```diff
@@ -0,0 +1,119 @@
+#
+# Cookbook:: fb_networkmanager
+# Recipe:: default
+#
+# Copyright (c) 2020-present, Vicarious, Inc.
+# Copyright (c) 2020-present, Facebook, Inc.
+# All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require './spec/spec_helper'
+require_relative '../libraries/default'
+
+describe FB::Networkmanager do
+  let(:mock_so) do
+    double('shellout')
+  end
+
+  before(:each) do
+    allow(mock_so).to receive(:run_command).and_return(mock_so)
+    allow(File).to receive(:exist?).with('/usr/bin/nmcli').
+      and_return(true)
+  end
+
+  context '#active_connections' do
+    it 'should return an empty hash if nmcli fails' do
+      expect(Mixlib::ShellOut).to receive(:new).and_return(mock_so)
+      allow(mock_so).to receive(:error?).and_return(true)
+      expect(FB::Networkmanager.active_connections).to eq({})
+    end
+
+    it 'should return an empty hash if nmcli has no active connections' do
+      expect(Mixlib::ShellOut).to receive(:new).and_return(mock_so)
+      allow(mock_so).to receive(:error?).and_return(false)
+      allow(mock_so).to receive(:stdout).and_return('')
+      expect(FB::Networkmanager.active_connections).to eq({})
+    end
+
+    it 'should return the parsed data correctly' do
+      expect(Mixlib::ShellOut).to receive(:new).and_return(mock_so)
+      allow(mock_so).to receive(:error?).and_return(false)
+      allow(mock_so).to receive(:stdout).and_return(
+        <<~EOF,
+          SomeNetwork:some-uuid:wifi:wlp2s0
+          AnotherNetwork:another-uid:eth:eth0
+          MyVPN:more-uuid:vpn:wlp2s0
+          tun0:all-the-uids:tun:tun0
+        EOF
+      )
+      expect(FB::Networkmanager.active_connections).to eq(
+        {
+          'SomeNetwork' => {
+            'uuid' => 'some-uuid',
+            'type' => 'wifi',
+            'device' => 'wlp2s0',
+          },
+          'AnotherNetwork' => {
+            'uuid' => 'another-uid',
+            'type' => 'eth',
+            'device' => 'eth0',
+          },
+          'MyVPN' => {
+            'uuid' => 'more-uuid',
+            'type' => 'vpn',
+            'device' => 'wlp2s0',
+          },
+          'tun0' => {
+            'uuid' => 'all-the-uids',
+            'type' => 'tun',
+            'device' => 'tun0',
+          },
+        },
+      )
+    end
+  end
+
+  # we don't test to make sure IniParse does what IniParse does,
+  # instead we teste the things we're doing custom...
+  context '#to_ini' do
```

Comment:
Add a test for array fields. If you call `FB::NetworkManager.to_ini` on this hash:

```
{
  'connection' => {
    'permissions' => %w{user:jgoguen:},
  },
  '802-1x' => {
    'identity' => 'jgoguen',
    'eap' => %w{tls},
  },
}
```

I expect this to be the generated config file:

```
[connection]
permissions=user:jgoguen:;

[802-1x]
identity=jgoguen
eap=tls;
```

`permissions` and `eap` are two fields that NetworkManager requires be multi-value, including the trailing semicolon after the last value.

Reply:
define doesn't work? it doesn't complain for sure, and it connects...

Reply:
In NM 1.20 and 1.22, what I found in my testing is that if I remove trailing semicolons I consistently have to restart NM for it to work with the changed connection file, but adding a trailing semicolon (which NM will do itself when it tries to update a multi-value field) works just by reloading the connections. So if we write out a connection file that removes a trailing semicolon, based on my experience we'll need to restart the entire NetworkManager daemon instead of just reloading the configs.

Reply:
I will experiment with this.